modules/gateway/peers_test.go

```diff
@@ -183,6 +183,231 @@ func TestConnect(t *testing.T) {
+			// Check that the error is the expected type.
+			if _, ok := err.(insufficientVersionError); !ok {
+				t.Fatalf("expected Connect to error with insufficientVersionError: %s", tt.msg)
+			}
+		} else {
+			// Check that the error is the expected error.
+			if err != tt.errWant {
+				t.Fatalf("expected Connect to error with '%v', but got '%v': %s", tt.errWant, err, tt.msg)
+			}
+		}
+		g.Disconnect(modules.NetAddress(listener.Addr().String()))
+	}
+	listener.Close()
+}
+
+// mockGatewayWithVersion is a mock implementation of Gateway that sends a mock
+// version on Connect instead of build.Version.
+type mockGatewayWithVersion struct {
+	*Gateway
+	version    string
+	versionACK chan string
+}
+
+// Connect is a mock implementation of modules.Gateway.Connect that provides a
+// mock version to peers it connects to instead of build.Version. The version
+// ack written by the remote peer is written to the versionACK channel.
+func (g mockGatewayWithVersion) Connect(addr modules.NetAddress) error {
+	conn, err := net.DialTimeout("tcp", string(addr), dialTimeout)
+	if err != nil {
+		return err
+	}
+	// send mocked version
+	if err := encoding.WriteObject(conn, g.version); err != nil {
+		return err
+	}
+	// read version ack
+	var remoteVersion string
+	if err := encoding.ReadObject(conn, &remoteVersion, maxAddrLength); err != nil {
+		return err
+	}
+	g.versionACK <- remoteVersion
+
+	return nil
+}
+
+// TestAcceptConnRejects tests that Gateway.acceptConn only accepts peers with
+// sufficient and valid versions.
+func TestAcceptConnRejects(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	g := newTestingGateway("TestAcceptConnRejects1", t)
+	defer g.Close()
+	mg := mockGatewayWithVersion{
+		Gateway:    newTestingGateway("TestAcceptConnRejects2", t),
+		versionACK: make(chan string),
+	}
+	defer mg.Close()
+
+	tests := []struct {
+		remoteVersion       string
+		versionResponseWant string
+		msg                 string
+	}{
+		// Test that acceptConn fails when the remote peer's version is "reject".
+		{
+			remoteVersion:       "reject",
+			versionResponseWant: "reject",
+			msg:                 "acceptConn shouldn't accept a remote peer whose version is \"reject\"",
+		},
+		// Test that acceptConn fails when the remote peer's version is ascii gibberish.
+		{
+			remoteVersion:       "foobar",
+			versionResponseWant: "reject",
+			msg:                 "acceptConn shouldn't accept a remote peer whose version is ascii giberish",
+		},
+		// Test that acceptConn fails when the remote peer's version is utf8 gibberish.
+		{
+			remoteVersion:       "世界",
+			versionResponseWant: "reject",
+			msg:                 "acceptConn shouldn't accept a remote peer whose version is utf8 giberish",
+		},
+		// Test that acceptConn fails when the remote peer's version is < 0.4.0 (0).
+		{
+			remoteVersion:       "0",
+			versionResponseWant: "reject",
+			msg:                 "acceptConn shouldn't accept a remote peer whose version is 0",
+		},
```

Comment:
probably good to have 00 and 0.0.0 and 00.00.00, I doubt any errors will turn up but often spamming 0's or 9's will get you something.